keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,294 @@
+for details.
+
+### Make the kube-apiserver aware of etcd's minor version
+
+This is only needed if we go with Alternative 2 (Use WatchProgressRequest to
+request watch updates when needed) since Alterative 1 is compatible with all
+etcd 3.x versions.
+
+Allow the etcd minor version to be specified in the kube-apiserver
+`--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  When client
+connections to etcd servers are established (and maybe periodically after that
+as well), check the current etcd version (via the etcd 'version' API) and warn
+the user if the version is older than the version of etcd provided in the
+flag.
+
+if `--storage-backend=etcd3` is provided, the minor etcd version will default to
+the detected etcd version minor version. This is for ease of use, since
+defaulting to `etcd3.0` would require additional configuration management by
+administrators. To make this safe, if the api-server gets an error when making a
+request that is enabled for newer etcd versions, it should warn in the logs that
+there might be an etcd version mismatch and fallback to the etcd functionality
+supported by all etcd 3.x versions.
+
+Also, Due to etcd upgrades and downgrades, there is no way to automatically
+detect the etcd version is a way that is guaranteed to be always correct, so the
+administrator must be able to set the version to a desired version.
+
+In addition to etcd minor version detection, all features requiring features
+introduced at a specific etcd minor version will have feature gates and will go
+through the usual kubernetes stability levels promotions (alpha, beta, GA).
+
+### Risks and Mitigations
+
+Alternative 2 can increase the number of round trips to etcd: One is required to
+get the latest revision, another might be needed to request the progress
+notify. Both of these requests are cheap and scalable, but we need to understand
+the latency impact of waiting for both of them before serving a response back to
+the client. If this turns out to have performance implications, it could be
+partially mitigated by configuring etcd to send regular progress events, which
+is something we will explore.
+
+In the worst case, the “wait until fresh” duration is exceeded, and then the
+request must be served from etcd anyway, which will be higher latency than
+immediately serving from etcd like we do today. We will need to set the
+“wait until fresh” duration to relatively short value (e.g. 20ms), which
+partially mitigates this.
+
+## Design Details
+
+### Pagination
+
+Because the watch cache does not paginate responses, clients that were
+previously getting pagination for RV="" requests will start getting unpaginated
+responses when this feature is enabled.
+
+The impacts of this can be explored when the feature is in alpha.
+
+### Test Plan
+
+We will need to scale and performance test this carefully. We’ll need to measure
+latency, throughput, volume of progress notify requests, and change to volume of
```

Comment:
I think this section should state how exactly you'll evaluate this? Measuring these things seems necessary but not sufficient?

Is this a correctness change or a performance change? Both?

Reply:
Like, I'm willing to pay latency to get throughput on startup + correctness. It'd be good to document how much we're willing to pay for how much benefit.

Reply:
Initially we're improving performance, primarily of LIST (with no RV provided), that doesn't change correctness. But then after that we're going to use instead of LIST (with RV=0) to improve correctness. We do expect a latency increase on those LIST calls but other wise no net loss to scale. I'll clarify with more detail in the doc.

@mm4tt and @wojtek-t have done some scale tests on etcd progress notify. I'll include more details here as well as what else we need to test.